// @nuxt/bridge-edge/dist/runtime/app.mjs
import { useHead } from "@unhead/vue";
import { getCurrentInstance } from "./composables.mjs";
export const isVue2 = true;
export const isVue3 = false;
export const defineNuxtComponent = function defineNuxtComponent2(...args) {
  const [options, key] = args;
  const { setup, head, ...opts } = options;
  if (!setup && !options.asyncData && !options.head) {
    return {
      ...options
    };
  }
  return {
    _fetchKeyBase: key,
    ...opts,
    setup(props, ctx) {
      const nuxtApp = useNuxtApp();
      const res = setup ? callWithNuxt(nuxtApp, setup, [props, ctx]) : {};
      if (options.head) {
        const nuxtApp2 = useNuxtApp();
        useHead(typeof options.head === "function" ? () => options.head(nuxtApp2) : options.head);
      }
      return res;
    }
  };
};
let currentNuxtAppInstance;
export const setNuxtAppInstance = (nuxt) => {
  currentNuxtAppInstance = nuxt;
};
export function callWithNuxt(nuxt, setup, args) {
  setNuxtAppInstance(nuxt);
  const p = args ? setup(...args) : setup();
  if (process.server) {
    setNuxtAppInstance(null);
  }
  return p;
}
export function defineNuxtPlugin(plugin) {
  return async (ctx, inject) => {
    const result = await callWithNuxt(ctx.$_nuxtApp, plugin, [ctx.$_nuxtApp]);
    if (result && result.provide) {
      for (const key in result.provide) {
        inject(key, result.provide[key]);
      }
    }
    return result;
  };
}
export const useNuxtApp = () => {
  const vm = getCurrentInstance();
  if (!vm) {
    if (!currentNuxtAppInstance) {
      throw new Error("nuxt app instance unavailable");
    }
    return currentNuxtAppInstance;
  }
  return vm.proxy.$_nuxtApp;
};
